cpp/include/cuspatial/experimental/detail/hausdorff.cuh

```diff
@@ -0,0 +1,178 @@
+
+#include <cuda/atomic>
+
+#include <type_traits>
+
+namespace cuspatial {
+
+namespace detail {
+
+template <typename T>
+constexpr auto magnitude_squared(T a, T b)
+{
+  return a * a + b * b;
+}
+
+/**
+ * @brief computes Hausdorff distance by equally dividing up work on a per-thread basis.
+ *
+ * Each thread is responsible for computing the distance from a single point in the input against
+ * all other points in the input. Because points in the input can originate from different spaces,
+ * each thread must know which spaces it is comparing. For the LHS argument, the point is always
+ * the same for any given thread and is determined once for that thread using a binary search of
+ * the provided space_offsets. Therefore if space 0 contains 10 points, the first 10 threads will
+ * know that the LHS space is 0. The 11th thread will know the LHS space is 1, and so on depending
+ * on the sizes/offsets of each space. Each thread then loops over each space, and uses an inner
+ * loop to loop over each point within that space, thereby knowing the RHS space and RHS point.
+ * the thread computes the minimum distance from it's LHS point to _any_ point in the RHS space, as
+ * this is the first step to computing Hausdorff distance. The second step of computing Hausdorff
+ * distance is to determine the maximum of these minimums, which is done by each thread writing
+ * it's minimum to the output using atomicMax. This is done once per thread per RHS space. Once
+ * all threads have run to completion, all "maximums of the minumum distances" (aka, directed
+ * Hausdorff distances) reside in the output.
+ *
+ * @tparam T type of coordinate, either float or double.
+ * @param num_points number of total points in the input (sum of points from all spaces)
+ * @param points x/y points to compute the distances between
+ * @param num_spaces number of spaces in the input
+ * @param space_offsets starting position of first point in each space
+ * @param results directed Hausdorff distances computed by kernel
+ */
+template <typename T,
+          typename Index,
+          typename PointsIter,
+          typename OffsetsIter,
+          typename OutputIter>
+__global__ void kernel_hausdorff(Index num_points,
+                                 PointsIter points,
+                                 Index num_spaces,
+                                 OffsetsIter space_offsets,
+                                 OutputIter results)
+{
+  using Point = typename std::iterator_traits<PointsIter>::value_type;
+
+  // determine the LHS point this thread is responsible for.
+  auto const thread_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  Index const lhs_p_idx = thread_idx;
+
+  if (lhs_p_idx >= num_points) { return; }
+
+  // determine the LHS space this point belongs to.
+  Index const lhs_space_idx =
+    thrust::distance(
+      space_offsets,
+      thrust::upper_bound(thrust::seq, space_offsets, space_offsets + num_spaces, lhs_p_idx)) -
+    1;
```

Comment:
Hmm, I thought I commented on this in the last review. Not sure why these are not added:

```suggestion
  // determine the LHS space this point belongs to.
  auto const lhs_space_iter = thrust::upper_bound(thrust::seq, space_offsets, space_offsets + num_spaces, lhs_p_idx));
  Index const lhs_space_idx = thrust::distance(space_offsets, thrust::prev(lhs_space_iter));
```

Reply:
This should make it more readable.